SU2_CFD/src/numerics_direct_heat.cpp

```diff
@@ -37,3 +37,318 @@
+    Mean_GradHeatVar[iVar] = new su2double [nDim];
+
+}
+
+CAvgGradCorrected_Heat::~CAvgGradCorrected_Heat(void) {
+
+  delete [] Edge_Vector;
+  delete [] Proj_Mean_GradHeatVar_Edge;
+  delete [] Proj_Mean_GradHeatVar_Kappa;
+  delete [] Proj_Mean_GradHeatVar_Corrected;
+  for (iVar = 0; iVar < nVar; iVar++)
+    delete [] Mean_GradHeatVar[iVar];
+  delete [] Mean_GradHeatVar;
+
+}
+
+void CAvgGradCorrected_Heat::ComputeResidual(su2double *val_residual, su2double **Jacobian_i, su2double **Jacobian_j, CConfig *config) {
+
+  AD::StartPreacc();
+  AD::SetPreaccIn(Coord_i, nDim); AD::SetPreaccIn(Coord_j, nDim);
+  AD::SetPreaccIn(Normal, nDim);
+  AD::SetPreaccIn(Temp_i); AD::SetPreaccIn(Temp_j);
+  AD::SetPreaccIn(ConsVar_Grad_i[0],nDim); AD::SetPreaccIn(ConsVar_Grad_j[0],nDim);
+  AD::SetPreaccIn(Thermal_Diffusivity_i); AD::SetPreaccIn(Thermal_Diffusivity_j);
+
+  Thermal_Diffusivity_Mean = 0.5*(Thermal_Diffusivity_i + Thermal_Diffusivity_j);
+
+  /*--- Compute vector going from iPoint to jPoint ---*/
+
+  dist_ij_2 = 0; proj_vector_ij = 0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    Edge_Vector[iDim] = Coord_j[iDim]-Coord_i[iDim];
+    dist_ij_2 += Edge_Vector[iDim]*Edge_Vector[iDim];
+    proj_vector_ij += Edge_Vector[iDim]*Normal[iDim];
+  }
+  if (dist_ij_2 == 0.0) proj_vector_ij = 0.0;
+  else proj_vector_ij = proj_vector_ij/dist_ij_2;
+
+  /*--- Mean gradient approximation. Projection of the mean gradient
+   in the direction of the edge ---*/
+
+  for (iVar = 0; iVar < nVar; iVar++) {
+    Proj_Mean_GradHeatVar_Edge[iVar] = 0.0;
+    Proj_Mean_GradHeatVar_Kappa[iVar] = 0.0;
+
+    for (iDim = 0; iDim < nDim; iDim++) {
+      Mean_GradHeatVar[iVar][iDim] = 0.5*(ConsVar_Grad_i[iVar][iDim] + ConsVar_Grad_j[iVar][iDim]);
+      Proj_Mean_GradHeatVar_Kappa[iVar] += Mean_GradHeatVar[iVar][iDim]*Normal[iDim];
+      Proj_Mean_GradHeatVar_Edge[iVar] += Mean_GradHeatVar[iVar][iDim]*Edge_Vector[iDim];
+    }
+    Proj_Mean_GradHeatVar_Corrected[iVar] = Proj_Mean_GradHeatVar_Kappa[iVar];
+    Proj_Mean_GradHeatVar_Corrected[iVar] -= Proj_Mean_GradHeatVar_Edge[iVar]*proj_vector_ij -
+    (Temp_j-Temp_i)*proj_vector_ij;
+  }
+
+  val_residual[0] = Thermal_Diffusivity_Mean*Proj_Mean_GradHeatVar_Corrected[0];
+
+  /*--- For Jacobians -> Use of TSL approx. to compute derivatives of the gradients ---*/
+
+  if (implicit) {
+    Jacobian_i[0][0] = -Thermal_Diffusivity_Mean*proj_vector_ij;
+    Jacobian_j[0][0] = Thermal_Diffusivity_Mean*proj_vector_ij;
+  }
+
+  AD::SetPreaccOut(val_residual, nVar);
+  AD::EndPreacc();
+}
+
+CSourceHeating::CSourceHeating(unsigned short val_nDim, unsigned short val_nVar, CConfig *config) : CNumerics(val_nDim, val_nVar, config) {
+
+}
+
+CSourceHeating::~CSourceHeating(void) { }
+
+void CSourceHeating::ComputeResidual(su2double *val_residual, CConfig *config) {
+  unsigned short iVar;
+
+  //su2double external_heat = config->GetExternalHeating();
+
+  //have to finish the implementation...
```

Comment:
Yep, we will want to clean this up so the user can specify volumetric heating with some access in the config file. Maybe we need to discuss a good design for this to make it easy for the user to specify the spatial bounds for volumetric heating. Any ideas?